File: unittest/trajectory/test_joint.cpp
#include <Eigen/Core>
#include <fstream>
#include <iomanip>
#include <iostream>
#include <ruckig/ruckig.hpp>

#include "trajgen.hpp"

using namespace ruckig;

void jointTrajgenTest1() {
  std::cout << "=== Test 1: JointTrajGen with NO constraints ===" << std::endl;

  const int ndof = 2;
  JointTrajGen<ndof> JointTraj(0.05);

  Eigen::VectorXd qPos0 = Eigen::VectorXd::Zero(ndof);
  Eigen::VectorXd qPos1 = Eigen::VectorXd::Zero(ndof);
  Eigen::VectorXd qUpperLimits = Eigen::VectorXd::Zero(ndof);
  Eigen::VectorXd qLowerLimits = Eigen::VectorXd::Zero(ndof);
  qPos0 << 0.3, -0.6;
  qPos1 << 2.0, -4.5;
  qUpperLimits.setConstant(3.0);
  qLowerLimits.setConstant(-3.0);

  Eigen::VectorXd qdMax(ndof), qddMax(ndof), qdddMax(ndof);
  qdMax << 1.0, 1.0;
  qddMax << 2.0, 1.0;
  qdddMax << 10.0, 5.0;

  JointTraj.setStartState(qPos0);
  JointTraj.setTargetState(qPos1);
  JointTraj.setJointPosLimits(qUpperLimits, qLowerLimits);
  JointTraj.setJointMotionLimits(qdMax, qddMax);

  std::cout << "From: [" << qPos0.transpose() << "] rad" << std::endl;
  std::cout << "To:   [" << qPos1.transpose() << "] rad" << std::endl;
  std::cout << "Motion limits - Vel: [" << qdMax.transpose() << "] rad/s" << std::endl;

  std::ofstream csv_file("joint1_data.csv");
  csv_file << "time";
  for (int i = 0; i < ndof; ++i) {
    csv_file << ",joint" << (i + 1);
  }
  for (int i = 0; i < ndof; ++i) {
    csv_file << ",vel" << (i + 1);
  }
  for (int i = 0; i < ndof; ++i) {
    csv_file << ",acc" << (i + 1);
  }
  csv_file << std::endl;

  std::cout << "\nExecuting trajectory (showing every 10th step):" << std::endl;
  std::cout << "t [s] | joint positions [rad] | joint velocities [rad/s]" << std::endl;
  std::cout << "------|----------------------|------------------------" << std::endl;

  int step = 0;
  Result result;
  do {
    result = JointTraj.update();

    // Print duration after first update
    if (step == 0) {
      double duration = JointTraj.getDuration();
      std::cout << "Trajectory duration: " << duration << "s" << std::endl;
    }

    auto qRef = JointTraj.getCurrPos();
    auto qdRef = JointTraj.getCurrVel();
    auto qddRef = JointTraj.getCurrAcc();
    double current_time = JointTraj.getTime();

    csv_file << std::fixed << std::setprecision(6) << current_time;
    for (int i = 0; i < ndof; ++i) {
      csv_file << "," << qRef(i);
    }
    for (int i = 0; i < ndof; ++i) {
      csv_file << "," << qdRef(i);
    }
    for (int i = 0; i < ndof; ++i) {
      csv_file << "," << qddRef(i);
    }
    csv_file << std::endl;

    if (step % 10 == 0) {
      std::cout << std::fixed << std::setprecision(3)
                << std::setw(5) << current_time << " | [";
      for (int i = 0; i < ndof; ++i) {
        std::cout << std::setw(6) << qRef(i) << (i < ndof - 1 ? "," : "");
      }
      std::cout << "] | [";
      for (int i = 0; i < ndof; ++i) {
        std::cout << std::setw(6) << qdRef(i) << (i < ndof - 1 ? "," : "");
      }
      std::cout << "]" << std::endl;
    }

    JointTraj.passToInput();
    step++;

  } while (result == Result::Working);

  csv_file.close();

  auto final_joints = JointTraj.getCurrPos();
  std::cout << "\nResult: " << static_cast<int>(result) << " (0=Working, 1=Finished, 2=Error)" << std::endl;
  std::cout << "Final joint positions: [" << final_joints.transpose() << "] rad" << std::endl;
  std::cout << "Total steps: " << step << std::endl;
  std::cout << "Data exported to: joint1_data.csv" << std::endl;
  std::cout << std::endl;
}

void jointTrajgenTest2() {
  std::cout << "=== Test 2: JointTrajGen with POSITION limits only ===" << std::endl;

  const int ndof = 6;
  JointTrajGen<ndof> JointTraj(0.001);

  Eigen::VectorXd qPos0 = Eigen::VectorXd::Zero(ndof);
  Eigen::VectorXd qPos1 = Eigen::VectorXd::Zero(ndof);
  Eigen::VectorXd qUpperLimits = Eigen::VectorXd::Zero(ndof);
  Eigen::VectorXd qLowerLimits = Eigen::VectorXd::Zero(ndof);
  qPos0 << 0.0, 0.0, 0.0, 1.0, -1.0, 2.0;
  qPos1 << 2.0, -2.5, 0.8, 4.0, -3.0, 6.0;

  qUpperLimits.setConstant(1.0);
  qLowerLimits.setConstant(-6.0);

  JointTraj.setStartState(qPos0);
  JointTraj.setTargetState(qPos1);
  JointTraj.setJointPosLimits(qUpperLimits, qLowerLimits);

  Eigen::VectorXd qdMax(ndof), qddMax(ndof), qdddMax(ndof);
  qdMax << 6.0, 4.0, 2.0, 8.0, 6.0, 10.0;       // rad/s
  qddMax << 3.0, 2.0, 1.0, 4.0, 3.0, 5.0;       // rad/s^2
  qdddMax << 15.0, 10.0, 5.0, 20.0, 15.0, 25.0; // rad/s^3
  // Test with velocity constraints only (should work with our fix)
  // JointTraj.setJointMotionLimits(qdMax);
  JointTraj.setJointMotionLimits(qdMax, qddMax);
  // JointTraj.setJointMotionLimits(qdMax, qddMax, qdddMax);

  std::cout << "From: [" << qPos0.transpose() << "] rad" << std::endl;
  std::cout << "To:   [" << qPos1.transpose() << "] rad (will be clamped)" << std::endl;
  std::cout << "Upper limits: [" << qUpperLimits.transpose() << "] rad" << std::endl;
  std::cout << "Lower limits: [" << qLowerLimits.transpose() << "] rad" << std::endl;

  std::ofstream csv_file("joint2_data.csv");
  csv_file << "time";
  for (int i = 0; i < ndof; ++i) {
    csv_file << ",joint" << (i + 1);
  }
  for (int i = 0; i < ndof; ++i) {
    csv_file << ",vel" << (i + 1);
  }
  for (int i = 0; i < ndof; ++i) {
    csv_file << ",acc" << (i + 1);
  }
  csv_file << std::endl;

  std::cout << "\nExecuting trajectory (showing every 10th step):" << std::endl;
  std::cout << "t [s] | joint positions [rad] | joint velocities [rad/s]" << std::endl;
  std::cout << "------|----------------------|------------------------" << std::endl;

  int step = 0;
  Result result;
  do {
    result = JointTraj.update();

    std::cout << "Update result: " << static_cast<int>(result) << " (0=Working, 1=Finished, 2=Error)" << std::endl;

    auto qRef = JointTraj.getCurrPos();
    auto qdRef = JointTraj.getCurrVel();
    auto qddRef = JointTraj.getCurrAcc();
    double current_time = JointTraj.getTime();

    std::cout << "Time: " << current_time << "s" << std::endl;

    csv_file << std::fixed << std::setprecision(6) << current_time;
    for (int i = 0; i < ndof; ++i) {
      csv_file << "," << qRef(i);
    }
    for (int i = 0; i < ndof; ++i) {
      csv_file << "," << qdRef(i);
    }
    for (int i = 0; i < ndof; ++i) {
      csv_file << "," << qddRef(i);
    }
    csv_file << std::endl;

    if (step % 10 == 0) {
      std::cout << std::fixed << std::setprecision(3)
                << std::setw(5) << current_time << " | [";
      for (int i = 0; i < ndof; ++i) {
        std::cout << std::setw(6) << qRef(i) << (i < ndof - 1 ? "," : "");
      }
      std::cout << "] | [";
      for (int i = 0; i < ndof; ++i) {
        std::cout << std::setw(6) << qdRef(i) << (i < ndof - 1 ? "," : "");
      }
      std::cout << "]" << std::endl;
    }

    JointTraj.passToInput();
    step++;

  } while (result == Result::Working);

  csv_file.close();

  auto final_joints = JointTraj.getCurrPos();
  std::cout << "\nFinal joint positions: [" << final_joints.transpose() << "] rad" << std::endl;
  std::cout << "Expected clamped to: ["
            << std::min(qPos1(0), qUpperLimits(0)) << ", "
            << std::max(qPos1(1), qLowerLimits(1)) << ", "
            << std::min(qPos1(2), qUpperLimits(2)) << "] rad" << std::endl;
  std::cout << "Total steps: " << step << std::endl;
  std::cout << "Data exported to: joint2_data.csv" << std::endl;
  std::cout << std::endl;
}

void jointTrajgenTest3() {
  std::cout << "=== Test 3: JointTrajGen with NON-ZERO boundary conditions ===" << std::endl;

  const int ndof = 3;
  JointTrajGen<ndof> JointTraj(0.001);

  Eigen::VectorXd qPos0(ndof), qVel0(ndof), qAcc0(ndof);
  Eigen::VectorXd qPos1(ndof), qVel1(ndof), qAcc1(ndof);

  qPos0 << 0.2, -0.3, 0.5;
  qVel0 << 0.1, -0.5, 0.2;
  qAcc0 << 0.05, 0.1, -0.1;

  qPos1 << 2.0, -4.5, 6.8;
  qVel1 << -0.2, 0.3, -0.1;
  qAcc1 << 0.1, -0.2, 0.15;

  Eigen::VectorXd qdMax(ndof), qddMax(ndof);
  qdMax.setConstant(2 * M_PI);
  qddMax.setConstant(M_PI);

  JointTraj.setStartState(qPos0, qVel0, qAcc0);
  JointTraj.setTargetState(qPos1, qVel1, qAcc1);
  JointTraj.setJointMotionLimits(qdMax, qddMax);

  std::cout << "From: pos=[" << qPos0.transpose() << "] rad" << std::endl;
  std::cout << "      vel=[" << qVel0.transpose() << "] rad/s" << std::endl;
  std::cout << "      acc=[" << qAcc0.transpose() << "] rad/s²" << std::endl;
  std::cout << "To:   pos=[" << qPos1.transpose() << "] rad" << std::endl;
  std::cout << "      vel=[" << qVel1.transpose() << "] rad/s" << std::endl;
  std::cout << "      acc=[" << qAcc1.transpose() << "] rad/s²" << std::endl;
  std::cout << "Duration: " << JointTraj.getDuration() << "s" << std::endl;

  std::ofstream csv_file("joint3_data.csv");
  csv_file << "time";
  for (int i = 0; i < ndof; ++i) {
    csv_file << ",joint" << (i + 1);
  }
  for (int i = 0; i < ndof; ++i) {
    csv_file << ",vel" << (i + 1);
  }
  for (int i = 0; i < ndof; ++i) {
    csv_file << ",acc" << (i + 1);
  }
  csv_file << std::endl;

  std::cout << "\nExecuting trajectory (showing every 15th step):" << std::endl;
  std::cout << "t [s] | joint positions [rad] | joint velocities [rad/s]" << std::endl;
  std::cout << "------|----------------------|------------------------" << std::endl;

  int step = 0;
  Result result;
  Eigen::VectorXd max_observed_vel = Eigen::VectorXd::Zero(ndof);

  do {
    result = JointTraj.update();

    auto qRef = JointTraj.getCurrPos();
    auto qdRef = JointTraj.getCurrVel();
    auto qddRef = JointTraj.getCurrAcc();
    double current_time = JointTraj.getTime();

    // Track maximum observed velocities
    for (int i = 0; i < ndof; ++i) {
      max_observed_vel(i) = std::max(max_observed_vel(i), std::abs(qdRef(i)));
    }

    csv_file << std::fixed << std::setprecision(6) << current_time;
    for (int i = 0; i < ndof; ++i) {
      csv_file << "," << qRef(i);
    }
    for (int i = 0; i < ndof; ++i) {
      csv_file << "," << qdRef(i);
    }
    for (int i = 0; i < ndof; ++i) {
      csv_file << "," << qddRef(i);
    }
    csv_file << std::endl;

    if (step % 15 == 0) {
      std::cout << std::fixed << std::setprecision(3)
                << std::setw(5) << current_time << " | [";
      for (int i = 0; i < ndof; ++i) {
        std::cout << std::setw(6) << qRef(i) << (i < ndof - 1 ? "," : "");
      }
      std::cout << "] | [";
      for (int i = 0; i < ndof; ++i) {
        std::cout << std::setw(6) << qdRef(i) << (i < ndof - 1 ? "," : "");
      }
      std::cout << "]" << std::endl;
    }

    JointTraj.passToInput();
    step++;

  } while (result == Result::Working);

  csv_file.close();

  auto final_joints = JointTraj.getCurrPos();
  auto final_vels = JointTraj.getCurrVel();
  auto final_accs = JointTraj.getCurrAcc();

  std::cout << "\nResult: " << static_cast<int>(result) << " (0=Working, 1=Finished, 2=Error)" << std::endl;
  std::cout << "Final pos: [" << final_joints.transpose() << "] rad" << std::endl;
  std::cout << "Final vel: [" << final_vels.transpose() << "] rad/s" << std::endl;
  std::cout << "Final acc: [" << final_accs.transpose() << "] rad/s²" << std::endl;
  std::cout << "Target pos:[" << qPos1.transpose() << "] rad" << std::endl;
  std::cout << "Target vel:[" << qVel1.transpose() << "] rad/s" << std::endl;
  std::cout << "Target acc:[" << qAcc1.transpose() << "] rad/s²" << std::endl;
  std::cout << "Max observed velocities: [" << max_observed_vel.transpose() << "] rad/s" << std::endl;
  std::cout << "Velocity limits were: [" << qdMax.transpose() << "] rad/s" << std::endl;
  std::cout << "Total steps: " << step << std::endl;
  std::cout << "Data exported to: joint3_data.csv" << std::endl;
  std::cout << std::endl;
}

int main() {
  std::cout << "=== JointTrajGen Comprehensive Testing (Dynamic Version) ===" << std::endl;
  std::cout << "Testing different constraint scenarios for joint trajectory generation\n"
            << std::endl;

  jointTrajgenTest1();

  jointTrajgenTest2();

  jointTrajgenTest3();

  std::cout << "=== All JointTrajGen tests completed ===" << std::endl;
  std::cout << "\nTest results exported as CSV files:" << std::endl;
  std::cout << "- joint1_data.csv" << std::endl;
  std::cout << "- joint2_data.csv" << std::endl;
  std::cout << "- joint3_data.csv" << std::endl;

  return 0;
}

File: python/src/RynnMotion/utils/trajectory/test_joint.py
#!/usr/bin/env python3
"""
Joint trajectory generation test with multiple scenarios.
Tests different constraint scenarios for joint trajectory generation.
"""

import numpy as np
from scipy.spatial.transform import Rotation as R
import csv
import math
import sys
import os

# Add the python directory to the path to import our modules
sys.path.append(os.path.join(os.path.dirname(__file__), '..', '..'))

from utils.trajectory.trajgen import JointTrajGen


def joint_trajgen_test1():
    """Test 1: JointTrajGen with NO constraints"""
    print("=== Test 1: JointTrajGen with NO constraints ===")
    
    ndof = 2
    joint_traj = JointTrajGen(ndof, 0.05)
    
    q_pos0 = np.array([0.3, -0.6])
    q_pos1 = np.array([2.0, -4.5])
    q_upper_limits = np.ones(ndof) * 3.0
    q_lower_limits = np.ones(ndof) * -3.0
    
    qd_max = np.array([1.0, 1.0])
    qdd_max = np.array([2.0, 1.0])
    qddd_max = np.array([10.0, 5.0])
    
    joint_traj.set_start_state(q_pos0)
    joint_traj.set_target_state(q_pos1)
    joint_traj.set_joint_pos_limits(q_upper_limits, q_lower_limits)
    joint_traj.set_joint_motion_limits(qd_max, qdd_max)
    
    print(f"From: [{', '.join(f'{x:.3f}' for x in q_pos0)}] rad")
    print(f"To:   [{', '.join(f'{x:.3f}' for x in q_pos1)}] rad")
    print(f"Motion limits - Vel: [{', '.join(f'{x:.3f}' for x in qd_max)}] rad/s")
    
    # Create CSV file for data export
    with open("joint1_data.csv", "w", newline="") as csv_file:
        writer = csv.writer(csv_file)
        # Write header
        header = ["time"]
        header.extend([f"joint{i+1}" for i in range(ndof)])
        header.extend([f"vel{i+1}" for i in range(ndof)])
        header.extend([f"acc{i+1}" for i in range(ndof)])
        writer.writerow(header)
        
        print("\nExecuting trajectory (showing every 10th step):")
        print("t [s] | joint positions [rad] | joint velocities [rad/s]")
        print("------|----------------------|------------------------")
        
        step = 0
        result = 0
        while result == 0:  # 0=Working
            result = joint_traj.update()
            
            # Print duration after first update
            if step == 0:
                duration = joint_traj.get_duration()
                print(f"Trajectory duration: {duration:.3f}s")
            
            q_ref = joint_traj.get_curr_pos()
            qd_ref = joint_traj.get_curr_vel()
            qdd_ref = joint_traj.get_curr_acc()
            current_time = joint_traj.get_time()
            
            # Write to CSV
            row = [f"{current_time:.6f}"]
            row.extend([f"{q:.6f}" for q in q_ref])
            row.extend([f"{qd:.6f}" for qd in qd_ref])
            row.extend([f"{qdd:.6f}" for qdd in qdd_ref])
            writer.writerow(row)
            
            if step % 10 == 0:
                pos_str = ", ".join(f"{q:6.3f}" for q in q_ref)
                vel_str = ", ".join(f"{qd:6.3f}" for qd in qd_ref)
                print(f"{current_time:5.3f} | [{pos_str}] | [{vel_str}]")
            
            joint_traj.pass_to_input()
            step += 1
            
            # Safety break to prevent infinite loops
            if step > 10000:
                print("Safety break: Maximum steps reached")
                break
    
    final_joints = joint_traj.get_curr_pos()
    print(f"\nResult: {result} (0=Working, 1=Finished, 2=Error)")
    print(f"Final joint positions: [{', '.join(f'{x:.3f}' for x in final_joints)}] rad")
    print(f"Total steps: {step}")
    print("Data exported to: joint1_data.csv")
    print()


def joint_trajgen_test2():
    """Test 2: JointTrajGen with POSITION limits only"""
    print("=== Test 2: JointTrajGen with POSITION limits only ===")
    
    ndof = 6
    joint_traj = JointTrajGen(ndof, 0.001)
    
    q_pos0 = np.array([0.0, 0.0, 0.0, 1.0, -1.0, 2.0])
    q_pos1 = np.array([2.0, -2.5, 0.8, 4.0, -3.0, 6.0])
    q_upper_limits = np.ones(ndof) * 1.0
    q_lower_limits = np.ones(ndof) * -6.0
    
    joint_traj.set_start_state(q_pos0)
    joint_traj.set_target_state(q_pos1)
    joint_traj.set_joint_pos_limits(q_upper_limits, q_lower_limits)
    
    qd_max = np.array([6.0, 4.0, 2.0, 8.0, 6.0, 10.0])       # rad/s
    qdd_max = np.array([3.0, 2.0, 1.0, 4.0, 3.0, 5.0])       # rad/s^2
    qddd_max = np.array([15.0, 10.0, 5.0, 20.0, 15.0, 25.0]) # rad/s^3
    
    # Test with velocity constraints only (should work with our fix)
    # joint_traj.set_joint_motion_limits(qd_max)
    joint_traj.set_joint_motion_limits(qd_max, qdd_max)
    # joint_traj.set_joint_motion_limits(qd_max, qdd_max, qddd_max)
    
    print(f"From: [{', '.join(f'{x:.3f}' for x in q_pos0)}] rad")
    print(f"To:   [{', '.join(f'{x:.3f}' for x in q_pos1)}] rad (will be clamped)")
    print(f"Upper limits: [{', '.join(f'{x:.3f}' for x in q_upper_limits)}] rad")
    print(f"Lower limits: [{', '.join(f'{x:.3f}' for x in q_lower_limits)}] rad")
    
    # Create CSV file for data export
    with open("joint2_data.csv", "w", newline="") as csv_file:
        writer = csv.writer(csv_file)
        # Write header
        header = ["time"]
        header.extend([f"joint{i+1}" for i in range(ndof)])
        header.extend([f"vel{i+1}" for i in range(ndof)])
        header.extend([f"acc{i+1}" for i in range(ndof)])
        writer.writerow(header)
        
        print("\nExecuting trajectory (showing every 10th step):")
        print("t [s] | joint positions [rad] | joint velocities [rad/s]")
        print("------|----------------------|------------------------")
        
        step = 0
        result = 0
        while result == 0:  # 0=Working
            result = joint_traj.update()
            
            print(f"Update result: {result} (0=Working, 1=Finished, 2=Error)")
            
            q_ref = joint_traj.get_curr_pos()
            qd_ref = joint_traj.get_curr_vel()
            qdd_ref = joint_traj.get_curr_acc()
            current_time = joint_traj.get_time()
            
            print(f"Time: {current_time:.6f}s")
            
            # Write to CSV
            row = [f"{current_time:.6f}"]
            row.extend([f"{q:.6f}" for q in q_ref])
            row.extend([f"{qd:.6f}" for qd in qd_ref])
            row.extend([f"{qdd:.6f}" for qdd in qdd_ref])
            writer.writerow(row)
            
            if step % 10 == 0:
                pos_str = ", ".join(f"{q:6.3f}" for q in q_ref)
                vel_str = ", ".join(f"{qd:6.3f}" for qd in qd_ref)
                print(f"{current_time:5.3f} | [{pos_str}] | [{vel_str}]")
            
            joint_traj.pass_to_input()
            step += 1
            
            # Safety break to prevent infinite loops
            if step > 10000:
                print("Safety break: Maximum steps reached")
                break
    
    final_joints = joint_traj.get_curr_pos()
    expected_clamped = [
        min(q_pos1[0], q_upper_limits[0]),
        max(q_pos1[1], q_lower_limits[1]),
        min(q_pos1[2], q_upper_limits[2])
    ]
    print(f"\nFinal joint positions: [{', '.join(f'{x:.3f}' for x in final_joints)}] rad")
    print(f"Expected clamped to: [{expected_clamped[0]:.3f}, {expected_clamped[1]:.3f}, {expected_clamped[2]:.3f}] rad")
    print(f"Total steps: {step}")
    print("Data exported to: joint2_data.csv")
    print()


def joint_trajgen_test3():
    """Test 3: JointTrajGen with NON-ZERO boundary conditions"""
    print("=== Test 3: JointTrajGen with NON-ZERO boundary conditions ===")
    
    ndof = 3
    joint_traj = JointTrajGen(ndof, 0.001)
    
    q_pos0 = np.array([0.2, -0.3, 0.5])
    q_vel0 = np.array([0.1, -0.5, 0.2])
    q_acc0 = np.array([0.05, 0.1, -0.1])
    
    q_pos1 = np.array([2.0, -4.5, 6.8])
    q_vel1 = np.array([-0.2, 0.3, -0.1])
    q_acc1 = np.array([0.1, -0.2, 0.15])
    
    qd_max = np.ones(ndof) * 2 * math.pi
    qdd_max = np.ones(ndof) * math.pi
    
    joint_traj.set_start_state(q_pos0, q_vel0, q_acc0)
    joint_traj.set_target_state(q_pos1, q_vel1, q_acc1)
    joint_traj.set_joint_motion_limits(qd_max, qdd_max)
    
    print(f"From: pos=[{', '.join(f'{x:.3f}' for x in q_pos0)}] rad")
    print(f"      vel=[{', '.join(f'{x:.3f}' for x in q_vel0)}] rad/s")
    print(f"      acc=[{', '.join(f'{x:.3f}' for x in q_acc0)}] rad/s²")
    print(f"To:   pos=[{', '.join(f'{x:.3f}' for x in q_pos1)}] rad")
    print(f"      vel=[{', '.join(f'{x:.3f}' for x in q_vel1)}] rad/s")
    print(f"      acc=[{', '.join(f'{x:.3f}' for x in q_acc1)}] rad/s²")
    print(f"Duration: {joint_traj.get_duration():.3f}s")
    
    # Create CSV file for data export
    with open("joint3_data.csv", "w", newline="") as csv_file:
        writer = csv.writer(csv_file)
        # Write header
        header = ["time"]
        header.extend([f"joint{i+1}" for i in range(ndof)])
        header.extend([f"vel{i+1}" for i in range(ndof)])
        header.extend([f"acc{i+1}" for i in range(ndof)])
        writer.writerow(header)
        
        print("\nExecuting trajectory (showing every 15th step):")
        print("t [s] | joint positions [rad] | joint velocities [rad/s]")
        print("------|----------------------|------------------------")
        
        step = 0
        result = 0
        max_observed_vel = np.zeros(ndof)
        
        while result == 0:  # 0=Working
            result = joint_traj.update()
            
            q_ref = joint_traj.get_curr_pos()
            qd_ref = joint_traj.get_curr_vel()
            qdd_ref = joint_traj.get_curr_acc()
            current_time = joint_traj.get_time()
            
            # Track maximum observed velocities
            for i in range(ndof):
                max_observed_vel[i] = max(max_observed_vel[i], abs(qd_ref[i]))
            
            # Write to CSV
            row = [f"{current_time:.6f}"]
            row.extend([f"{q:.6f}" for q in q_ref])
            row.extend([f"{qd:.6f}" for qd in qd_ref])
            row.extend([f"{qdd:.6f}" for qdd in qdd_ref])
            writer.writerow(row)
            
            if step % 15 == 0:
                pos_str = ", ".join(f"{q:6.3f}" for q in q_ref)
                vel_str = ", ".join(f"{qd:6.3f}" for qd in qd_ref)
                print(f"{current_time:5.3f} | [{pos_str}] | [{vel_str}]")
            
            joint_traj.pass_to_input()
            step += 1
            
            # Safety break to prevent infinite loops
            if step > 10000:
                print("Safety break: Maximum steps reached")
                break
    
    final_joints = joint_traj.get_curr_pos()
    final_vels = joint_traj.get_curr_vel()
    final_accs = joint_traj.get_curr_acc()
    
    print(f"\nResult: {result} (0=Working, 1=Finished, 2=Error)")
    print(f"Final pos: [{', '.join(f'{x:.3f}' for x in final_joints)}] rad")
    print(f"Final vel: [{', '.join(f'{x:.3f}' for x in final_vels)}] rad/s")
    print(f"Final acc: [{', '.join(f'{x:.3f}' for x in final_accs)}] rad/s²")
    print(f"Target pos:[{', '.join(f'{x:.3f}' for x in q_pos1)}] rad")
    print(f"Target vel:[{', '.join(f'{x:.3f}' for x in q_vel1)}] rad/s")
    print(f"Target acc:[{', '.join(f'{x:.3f}' for x in q_acc1)}] rad/s²")
    print(f"Max observed velocities: [{', '.join(f'{x:.3f}' for x in max_observed_vel)}] rad/s")
    print(f"Velocity limits were: [{', '.join(f'{x:.3f}' for x in qd_max)}] rad/s")
    print(f"Total steps: {step}")
    print("Data exported to: joint3_data.csv")
    print()


def main():
    """Main function to run all joint trajectory tests."""
    print("=== JointTrajGen Comprehensive Testing (Dynamic Version) ===")
    print("Testing different constraint scenarios for joint trajectory generation\n")
    
    joint_trajgen_test1()
    joint_trajgen_test2()
    joint_trajgen_test3()
    
    print("=== All JointTrajGen tests completed ===")
    print("\nTest results exported as CSV files:")
    print("- joint1_data.csv")
    print("- joint2_data.csv")
    print("- joint3_data.csv")


if __name__ == "__main__":
    main()